local-cli/generator/copyProjectTemplateAndReplace.js

```diff
@@ -0,0 +1,95 @@
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @flow
+ */
+'use strict';
+
+const chalk = require('chalk');
+const copyAndReplace = require('../util/copyAndReplace');
+const fs = require('fs');
+const path = require('path');
+const prompt = require('./promptSync')();
+const walk = require('../util/walk');
+
+/**
+ * Util for creating a new React Native project.
+ * Copy the project from a template and use the correct project name in
+ * all files.
+ * @param srcPath e.g. '/Users/martin/AwesomeApp/node_modules/react-native/local-cli/templates/HelloWorld'
+ * @param destPath e.g. '/Users/martin/AwesomeApp'
+ * @param newProjectName e.g. 'AwesomeApp'
+ */
+function copyProjectTemplateAndReplace(srcPath, destPath, newProjectName, options) {
+  if (!srcPath) throw new Error('Need a path to copy from');
+  if (!destPath) throw new Error('Need a path to copy to');
+  if (!newProjectName) throw new Error('Need a project name');
+
+  walk(srcPath).forEach(absoluteSrcFilePath => {
+
+    // 'react-native upgrade'
+    if (options && options.upgrade) {
+      // Don't upgrade these files
+      const fileName = path.basename(absoluteSrcFilePath);
+      // TODO __tests__/index.ios.js
+      // TODO __tests__/index.android.js
+      if (fileName === 'index.ios.js') return;
+      if (fileName === 'index.android.js') return;
+    }
+
+    const relativeFilePath = path.relative(srcPath, absoluteSrcFilePath);
+    const relativeRenamedPath = relativeFilePath
+      .replace(/HelloWorld/g, newProjectName)
+      .replace(/helloworld/g, newProjectName.toLowerCase());
+    
+    copyAndReplace(
+      absoluteSrcFilePath,
+      path.resolve(destPath, relativeRenamedPath),
+      {
+        'HelloWorld': newProjectName,
+        'helloworld': newProjectName.toLowerCase(),
+      },
+      options && options.upgrade ? (_, contentChanged) => {
+        return upgradeFileContentChangedCallback(absoluteSrcFilePath, relativeRenamedPath, contentChanged)
+      } : null
+    );
+  });
+}
+
+function upgradeFileContentChangedCallback(
+  absoluteSrcFilePath,
+  relativeDestPath,
+  contentChanged
+) {
+  if (contentChanged === 'new') {
+    console.log(chalk.bold('new') + ' ' + relativeDestPath);
+    return 'overwrite';
+  } else if (contentChanged === 'changed') {
+    console.log(chalk.bold(relativeDestPath) + ' ' +
+      'has changed in the new version.\nDo you want to keep your ' + 
+      relativeDestPath + ' or replace it with the ' +
+      'latest version?\nIf you ever made any changes ' +
+      'to this file, you\'ll probably want to keep it.\n' +
+      'You can see the new version here: ' + absoluteSrcFilePath + '\n' +
+      'Please answer "keep" / "replace" for ' + relativeDestPath + ':');
+    const answer = prompt();
+    if (answer == 'replace') {
```

Comment:
If the user makes a typo in "replace", the file is kept. It could lead to a lot of madness :p